models/POP/shell_scripts/cesm2_1/perfect_model.csh

```diff
@@ -0,0 +1,147 @@
+set FILE = `head -n 1 rpointer.ocn.restart`
+# :t is a csh modifier that returns the "tail" of the filename
+# essentially stripping away the filepath that prepends the file
+set FILE = $FILE:t
+# :r is a csh modifier that returns the "root" of the filename 
+# essentially stripping away the file suffix
+set FILE = $FILE:r
+# :e is a csh modifier that returns the "extension" of the filename
+# since we stripped away the .nc in the previous command, the current
+# extension is the datetime in the form YYYY-MM-DD-SSSSS
+set OCN_DATE_EXT = `echo $FILE:e`
+# Piping the datetime to this sed command replaces the hyphens with spaces
+set OCN_DATE     = `echo $FILE:e | sed -e "s#-# #g"`
+# Piping the indexed substrings to bc strips away the leading zeros
+set OCN_YEAR     = `echo $OCN_DATE[1] | bc`
+set OCN_MONTH    = `echo $OCN_DATE[2] | bc`
+set OCN_DAY      = `echo $OCN_DATE[3] | bc`
+set OCN_SECONDS  = `echo $OCN_DATE[4] | bc`
+set OCN_HOUR     = `echo $OCN_DATE[4] / 3600 | bc`
+
+echo "valid time of model is $OCN_YEAR $OCN_MONTH $OCN_DAY $OCN_SECONDS (seconds)"
+echo "valid time of model is $OCN_YEAR $OCN_MONTH $OCN_DAY $OCN_HOUR (hours)"
+
+#=========================================================================
+# Block 2: Populate a run-time directory with the obs_seq needed to run DART.
+#=========================================================================
+
+# Get observation sequence file ... or die right away.
+# The observation file names have a time that matches the stopping time of POP.
+
+set YYYYMM   = `printf %04d%02d                ${OCN_YEAR} ${OCN_MONTH}`
+set OBSFNAME = `printf obs_seq.0Z.%04d%02d%02d ${OCN_YEAR} ${OCN_MONTH} ${OCN_DAY}`
+set OBS_FILE = ${BASEOBSDIR}/${YYYYMM}/${OBSFNAME}
+
+if (  -e   ${OBS_FILE} ) then
+   ${LINK} ${OBS_FILE} obs_seq.in
+else
+   echo "ERROR ... no observation file $OBS_FILE"
+   echo "ERROR ... no observation file $OBS_FILE"
+   exit -2
+endif
+
+#=========================================================================
+# Block 3: Populate a run-time directory with the input.nml needed to run DART.
+#=========================================================================
+
+echo "`date` -- BEGIN COPY BLOCK"
+
+if ( -e ${CASEROOT}/input.nml ) then
+   ${COPY} ${CASEROOT}/input.nml .
+else
+   echo "ERROR ... DART required file ${CASEROOT}/input.nml not found ... ERROR"
+   echo "ERROR ... DART required file ${CASEROOT}/input.nml not found ... ERROR"
+   exit -3
+endif
+
+echo "`date` -- END COPY BLOCK"
+
+set OCN_RESTART_FILENAME = ${CASE}.pop.r.${OCN_DATE_EXT}.nc
+# BKJ I think setting this is unnecessary
+# set     OCN_NML_FILENAME = pop_in
+
+${LINK} ./$OCN_RESTART_FILENAME pop.r.nc
+# BKJ I think setting this is unnecessary
+# ${LINK} ../$OCN_NML_FILENAME     pop_in
+
+#=========================================================================
+# Block 4: Advance the model and harvest the synthetic observations.
+# output files are:
+# True_state.nc   ...... the DART state
```

Comment:
Is the creation of this still 'automatic'?  Does it depend on a namelist setting?
Is the file 'True_state.nc' or 'true_state.nc' or ... does it have a stage name ?

Reply:
Unless the input.nml:perfect_model_obs_nml:  write_output_state_to_file = .true.    there is no output. The input.nml in the repo has this as .false.

Reply:
More information:
When `write_output_state_to_file = .true.` the `perfect_restart` is created - as a netCDF file. The input.nml could/should have a .nc on this filename.
When `single_file_out = .true.`, the `true_state.nc` file is created. (the input.nml has this as `.false.` - not sure what you want as the default action.


```
   write_output_state_to_file = .true.
   single_file_out            = .true.
   output_state_files         = 'perfect_restart'
```